### ISSUE TITLE
 Option 'z_offset' in section 'bltouch' must be specified 

### ISSUE THREAD

**Issue by Talha909** [Klipper3d/klipper#4970]
I did the z calibration and after that I click on accept and then save_config. After pressing the save config I got this error message 

Option 'z_offset' in section 'bltouch' must be specified

Once the underlying issue is corrected, use the "RESTART"
command to reload the config and restart the host software.
Printer is halted
![zoffset](https://user-images.githubusercontent.com/29024390/143771030-291ad6e0-9269-4f03-a995-8f449c13f5ad.JPG)
[klippy.log](https://github.com/Klipper3d/klipper/files/7613668/klippy.log)
[printer (2).txt](https://github.com/Klipper3d/klipper/files/7613670/printer.2.txt)

**Comment by gnutison** [comment]
add to [bltouch] z_offset: 0

**Comment by Talha909** [comment]
I have set the Z_offset:0 but when i calibrate the proble. I do testz and then accept and then save_config.  At that time klipper commented z offset and through the error message. I then open the printer.cfg file and uncommented it and save it. The new value is saved in the file. But the offset did not work.  There is always gap between the nozzle and bed.

**Comment by gnutison** [comment]
I had exactly issues like you before.
Adding z_offset: 0 in bltouch was "little backdoor" to avoid an error message so I can move to next steps PROBE_CALIBRATE. After properly Probe Calibrate  z offset should be commented in bltouch section and the real one is down below when u scrolled down. Under this text
#*# <---------------------- SAVE_CONFIG ---------------------->
#*# DO NOT EDIT THIS BLOCK OR BELOW. The contents are auto-generated.

**Comment by snafu1282** [comment]
I'm have the same issue, except Klipper will not accept the fact that there is a SAVE_CONFIG section. In fact, I wind up with 2 SAVE_CONFIG sections, and I still get the "Option 'z_offset' in section 'bltouch' must be specified" error.

# This file contains common pin mappings for the BIGTREETECH SKR V1.4
# board. To use this config, the firmware should be compiled for the
# LPC1768 or LPC1769(Turbo).

# See docs/Config_Reference.md for a description of parameters.

[bltouch]
sensor_pin: P0.10  # Pull-up (^ symbol) needed in open drain mode
control_pin: P2.0
# Some BLTouch V3 and many clones apparently require this, though mine didnt:
pin_up_touch_mode_reports_triggered: true
#flavor: genuine_smart_3.1
x_offset: -43
y_offset: -10
# z_offset = 1.199
pin_move_time: 1
samples: 2

[stepper_x]
step_pin: P2.2
dir_pin: !P2.6
enable_pin: !P2.1
microsteps: 16
rotation_distance: 80
#endstop_pin: !P1.29
endstop_pin: P1.29
position_endstop: 228
position_max: 228
homing_speed: 40
position_min: 0

[stepper_y]
step_pin: P0.19
dir_pin: !P0.20
enable_pin: !P2.8
microsteps: 16
rotation_distance: 80
#endstop_pin: !P1.28
endstop_pin: P1.28
position_endstop: 220
position_max: 220
homing_speed: 40
position_min: 0

[stepper_z]
step_pin: P0.22
dir_pin: !P2.11
#dir_pin: P2.11
enable_pin: !P0.21
microsteps: 16
rotation_distance: 8
#endstop_pin: !P1.27
#endstop_pin: P1.28
#position_endstop: 0.0
position_max: 300
endstop_pin: probe:z_virtual_endstop
position_min: -3

[extruder]
step_pin: P2.13
#dir_pin: !P0.11
dir_pin: P0.11
enable_pin: !P2.12
microsteps: 16
max_extrude_only_distance: 100.0
#rotation_distance: 22.8346860205898
#rotation_distance: 44
#rotation_distance: 22.8346860205898
rotation_distance: 45.6693720411796
nozzle_diameter: 0.400
filament_diameter: 1.750
heater_pin: P2.7
sensor_type: EPCOS 100K B57560G104F
sensor_pin: P0.24
#sensor_pin: P0.23
control: pid
pid_Kp: 21.527
pid_Ki: 1.063
pid_Kd: 108.982
min_temp: 0
max_temp: 280
#min_extrude_temp: 0
pressure_advance = 0.135

[heater_bed]
heater_pin: P2.5
sensor_type: EPCOS 100K B57560G104F
sensor_pin: P0.25
control: pid
pid_Kp: 54.027
pid_Ki: 0.770
pid_Kd: 948.182
min_temp: 0
max_temp: 130

[fan]
pin: P2.3

[mcu]
serial: /dev/serial/by-id/usb-Klipper_lpc1769_13E0FF0FA69869AF364A415EC52000F5-if00
#serial: /dev/serial/by-id/usb-Klipper_Klipper_firmware_12345-if00
restart_method: command

[printer]
kinematics: cartesian
max_velocity: 500
max_accel: 2000
max_z_velocity: 10
max_z_accel: 100
max_accel_to_decel: 1000

[display]
lcd_type: st7920
cs_pin: EXP1_7
sclk_pin: EXP1_6
sid_pin: EXP1_8
encoder_pins: ^EXP1_5, ^EXP1_3
click_pin: ^!EXP1_2

[virtual_sdcard]
path: ~/gcode_files

[display_status]

[pause_resume]

###############################################
##  Probing/Mesh
###############################################



[safe_z_home]
home_xy_position: 160.5,137.5
speed: 50.0
z_hop: 10.0
z_hop_speed: 5.0

[bed_mesh]
speed: 30
horizontal_move_z: 10
mesh_min: 20,18
mesh_max: 180,206
probe_count: 5,5
fade_start: 1.0
mesh_pps: 2,2

[screws_tilt_adjust]
screw1: 66,47
screw1_name: vorne links
screw2: 225,47
screw2_name: vorne rechts
screw3: 225,216
screw3_name: hinten rechts
screw4: 66,216
screw4_name: hinten links
horizontal_move_z: 10.
speed: 50.
screw_thread: CW-M4

[mcu rpi]
serial: /tmp/klipper_host_mcu

[adxl345]
cs_pin: rpi:None

[resonance_tester]
accel_chip: adxl345
probe_points: 100,100,20 # an example

#[input_shaper]
#shaper_freq_x: 49
#shaper_type_x: mzv
#shaper_freq_y: 74.8
#shaper_type_y: 2hump_ei
[input_shaper]
shaper_freq_x: 93
shaper_type_x: 3hump_ei
shaper_freq_y: 49
shaper_type_y: 2hump_ei

#[extruder1]
#step_pin: P1.15
#dir_pin: P1.14
#enable_pin: !P1.16
#heater_pin: P2.4
#sensor_pin: P0.23
#...

[gcode_macro PAUSE]
description: Pause the actual running print
rename_existing: PAUSE_BASE
# change this if you need more or less extrusion
variable_extrude: 1.0
gcode:
  ##### read E from pause macro #####
  {% set E = printer["gcode_macro PAUSE"].extrude|float %}
  ##### set park positon for x and y #####
  # default is your max posion from your printer.cfg
  {% set x_park = printer.toolhead.axis_maximum.x|float - 5.0 %}
  {% set y_park = printer.toolhead.axis_maximum.y|float - 5.0 %}
  ##### calculate save lift position #####
  {% set max_z = printer.toolhead.axis_maximum.z|float %}
  {% set act_z = printer.toolhead.position.z|float %}
  {% if act_z < (max_z - 2.0) %}
      {% set z_safe = 2.0 %}
  {% else %}
      {% set z_safe = max_z - act_z %}
  {% endif %}
  ##### end of definitions #####
  PAUSE_BASE
  G91
  {% if printer.extruder.can_extrude|lower == 'true' %}
    G1 E-{E} F2100
  {% else %}
    {action_respond_info("Extruder not hot enough")}
  {% endif %}
  {% if "xyz" in printer.toolhead.homed_axes %}
    G1 Z{z_safe} F900
    G90
    G1 X{x_park} Y{y_park} F6000
  {% else %}
    {action_respond_info("Printer not homed")}
  {% endif %} 

#RESUME
[gcode_macro RESUME]
description: Resume the actual running print
rename_existing: RESUME_BASE
gcode:
  ##### read E from pause macro #####
  {% set E = printer["gcode_macro PAUSE"].extrude|float %}
  #### get VELOCITY parameter if specified ####
  {% if 'VELOCITY' in params|upper %}
    {% set get_params = ('VELOCITY=' + params.VELOCITY)  %}
  {%else %}
    {% set get_params = "" %}
  {% endif %}
  ##### end of definitions #####
  {% if printer.extruder.can_extrude|lower == 'true' %}
    G91
    G1 E{E} F2100
  {% else %}
    {action_respond_info("Extruder not hot enough")}
  {% endif %}  
  RESUME_BASE {get_params}

#CANCEL_PRINT
[gcode_macro CANCEL_PRINT]
description: Cancel the actual running print
rename_existing: CANCEL_PRINT_BASE
gcode:
  TURN_OFF_HEATERS
  CANCEL_PRINT_BASE

######################################################################
# 128x64 Full Graphic Creality CR10 / ENDER 3 stockdisplay
######################################################################

# This section is used for a Creality "12864" display with a single
# ribbon cable between the display's EXP3 plug and the
# micro-controller board's EXP1 connector.

########################################
# EXP1 / EXP2 (display) pins
########################################
[board_pins]
aliases:
    # EXP1 header
    EXP1_1=P1.30, EXP1_3=P1.18, EXP1_5=P1.20, EXP1_7=P1.22, EXP1_9=<GND>,
    EXP1_2=P0.28, EXP1_4=P1.19, EXP1_6=P1.21, EXP1_8=P1.23, EXP1_10=<5V>,
    # EXP2 header
    EXP2_1=P0.17, EXP2_3=P3.26, EXP2_5=P3.25, EXP2_7=P1.31, EXP2_9=<GND>,
    EXP2_2=P0.15, EXP2_4=P0.16, EXP2_6=P0.18, EXP2_8=<RST>, EXP2_10=<NC>
    # Pins EXP2_1, EXP2_6, EXP2_2 are also MISO, MOSI, SCK of bus "ssp0"

# See the sample-lcd.cfg file for definitions of common LCD displays.

# Set z offset = -1.001 after testz=-4

#*# <---------------------- SAVE_CONFIG ---------------------->
#*# DO NOT EDIT THIS BLOCK OR BELOW. The contents are auto-generated.
#*#
#*# [bltouch]

#*# z_offset = 1.199

#*# [bed_mesh default]
#*# version = 1
#*# points =
#*# 	0.010000, -0.006250, 0.048750, 0.063750, 0.041250
#*# 	-0.068750, -0.045000, -0.003750, -0.015000, 0.006250
#*# 	-0.070000, -0.073750, -0.035000, -0.015000, 0.015000
#*# 	-0.072500, -0.063750, -0.026250, -0.026250, 0.046250
#*# 	-0.032500, -0.010000, -0.048750, -0.045000, 0.037500
#*# tension = 0.2
#*# min_x = 20.0
#*# algo = lagrange
#*# y_count = 5
#*# mesh_y_pps = 2
#*# min_y = 18.0
#*# x_count = 5
#*# max_y = 206.0
#*# mesh_x_pps = 2
#*# max_x = 180.0

#*# <---------------------- SAVE_CONFIG ---------------------->
#*# DO NOT EDIT THIS BLOCK OR BELOW. The contents are auto-generated.
#*#
#*# [bltouch]
#*# z_offset = 1.399

**Comment by snafu1282** [comment]
I'm have the same issue, except Klipper will not accept the fact that there is a SAVE_CONFIG section. In fact, I wind up with 2 SAVE_CONFIG sections, and I still get the "Option 'z_offset' in section 'bltouch' must be specified" error.

Here's the relevant info. I have also attached a zip of printer.cfg.

```
[bltouch]
sensor_pin: P0.10  # Pull-up (^ symbol) needed in open drain mode
control_pin: P2.0
# Some BLTouch V3 and many clones apparently require this, though mine didnt:
pin_up_touch_mode_reports_triggered: true
#flavor: genuine_smart_3.1
x_offset: -43
y_offset: -10
# z_offset = 1.199
pin_move_time: 1
samples: 2

# Set z offset = -1.001 after testz=-4

#*# <---------------------- SAVE_CONFIG ---------------------->
#*# DO NOT EDIT THIS BLOCK OR BELOW. The contents are auto-generated.
#*#
#*# [bltouch]

#*# z_offset = 1.199

#*# [bed_mesh default]
#*# version = 1
#*# points =
#*# 	0.010000, -0.006250, 0.048750, 0.063750, 0.041250
#*# 	-0.068750, -0.045000, -0.003750, -0.015000, 0.006250
#*# 	-0.070000, -0.073750, -0.035000, -0.015000, 0.015000
#*# 	-0.072500, -0.063750, -0.026250, -0.026250, 0.046250
#*# 	-0.032500, -0.010000, -0.048750, -0.045000, 0.037500
#*# tension = 0.2
#*# min_x = 20.0
#*# algo = lagrange
#*# y_count = 5
#*# mesh_y_pps = 2
#*# min_y = 18.0
#*# x_count = 5
#*# max_y = 206.0
#*# mesh_x_pps = 2
#*# max_x = 180.0

#*# <---------------------- SAVE_CONFIG ---------------------->
#*# DO NOT EDIT THIS BLOCK OR BELOW. The contents are auto-generated.
#*#
#*# [bltouch]
#*# z_offset = 1.399
```
[printer2021dec19 #2.zip](https://github.com/Klipper3d/klipper/files/7741676/printer2021dec19.2.zip)

**Comment by Sineos** [comment]
> So this begs the question: why 2 SAVE_CONFIGs, and which 1 takes precedence?

Your excerpt looks like you have manually tinkered with the `SAVE_CONFIG` section as it does not produce blank lines. I would delete the whole block and start over.

**Comment by snafu1282** [comment]
> > So this begs the question: why 2 SAVE_CONFIGs, and which 1 takes precedence?
> 
> Your excerpt looks like you have manually tinkered with the `SAVE_CONFIG` section as it does not produce blank lines. I would delete the whole block and start over.

Oh how I wish that was true. It would be so easy to fix in that case. Unfortunately, it's not the case. The 2 sections shown were both completely auto-generated. That was after doing what you suggested: i.e. completely deleting the SAVE_CONFIG section.

I wonder what would happen if I deleted the blank lines?

**Comment by snafu1282** [comment]
> I wonder what would happen if I deleted the blank lines?

So I have my answer. I removed the blank lines from the top SAVE_CONFIG and deleted the bottom 2 (yes, it had gone up to 3 SAVE_CONFIGs after doing another probe calibration). I also ensured that there was no space between the # and z_offset in [bltouch]. After a restart, there was no error. At the moment, it's working. Fantastic.

Thank you for the info about blank lines: not something I'm likely to forget any time soon 👍
